### ISSUE TITLE
Authorization FallbackPolicy - requirements not evaluated

### ISSUE THREAD

**Issue by JanKotschenreuther** [dotnet/aspnetcore#48019]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

The requirements of a `AuthorizationPolicy` are not evaluated, if the policy is used as FallbackPolicy.
The problem can be reproduced with a WebApp project template.

I added the following services.

```csharp
services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
    .AddJwtBearer(JwtBearerDefaults.AuthenticationScheme, options =>
    {
        options.TokenValidationParameters = new TokenValidationParameters()
        {
            ValidateIssuerSigningKey = true,
            ValidateIssuer = true,
            ValidateAudience = true,
            ValidIssuer = "localhost",
            ValidAudience = "localhost",
            IssuerSigningKey = GetSymmetricSecurityKey(),
        };
    });

services.AddAuthorization(options =>
{
    var basePolicy = new AuthorizationPolicyBuilder(JwtBearerDefaults.AuthenticationScheme)
        .RequireAuthenticatedUser()
        .RequireClaim("SomeClaim_1")
        .Build();

    var defaultPolicy = new AuthorizationPolicyBuilder(basePolicy)
        .RequireClaim("SomeClaim_2")
        .Build();

    options.FallbackPolicy = defaultPolicy;

    options.AddPolicy("Default", defaultPolicy);
});
```

And I added the following middleware configuration.

```csharp
app.UseAuthentication();
app.UseAuthorization();

app.UseEndpoints(endpoints =>
{
    endpoints.MapControllers().RequireAuthorization();
});
```

And I added the following middleware configuration.

```csharp
services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
    .AddJwtBearer(JwtBearerDefaults.AuthenticationScheme, options =>
    {
        options.TokenValidationParameters = new TokenValidationParameters()
        {
            ValidateIssuerSigningKey = true,
            ValidateIssuer = true,
            ValidateAudience = true,
            ValidIssuer = "localhost",
            ValidAudience = "localhost",
            IssuerSigningKey = GetSymmetricSecurityKey(),
        };
    });

services.AddAuthorization(options =>
{
    var basePolicy = new AuthorizationPolicyBuilder(JwtBearerDefaults.AuthenticationScheme)
        .RequireAuthenticatedUser()
        .RequireClaim("SomeClaim_1")
        .Build();

    var defaultPolicy = new AuthorizationPolicyBuilder(basePolicy)
        .RequireClaim("SomeClaim_2")
        .Build();

    options.FallbackPolicy = defaultPolicy;

    options.AddPolicy("Default", defaultPolicy);
});
```

I also added a single controller.

```csharp
    [Route("api/[controller]")]
    [ApiController]
    public class HelloWorldController : ControllerBase
    {
        [AllowAnonymous]
        [HttpGet("First")]
        public IActionResult First()
        {
            return Ok("1. Hallo Welt");
        }

        [HttpGet("Second")]
        public IActionResult Second()
        {
            return Ok("2. Hallo Welt");
        }

        [Authorize("Default")]
        [HttpGet("Third")]
        public IActionResult Third()
        {
            return Ok("3. Hallo Welt");
        }
    }
```

To reproduce the problem, I am using the following payload for my JWT.
The claim `SomeClaim_2` is configured as required on the defaultPolicy which is used on the FallbackPolicy.

```json
{
  "SomeClaim_1": "SomeValue_1",
  "nbf": 1683019121,
  "exp": 1683105521,
  "iat": 1683019121,
  "iss": "localhost",
  "aud": "localhost"
}
```

The endpoint `Second` returns a 200 OK, even though the claim `SomeClaim_2` is required.
The endpoint `Third` returns a  403 Forbidden, as expected, because the claim `SomeClaim_2` is required.

On a real project I also observed that implementations of `AuthorizationHandler<TRequirement>` are not fired or getting called if the FallbackPolicy is used.
Using the FallbackPolicy as named policy, they also get called as expected.

### Expected Behavior

I expected both endpoints `Second` and `Third` to return the same result 403 Forbidden.

Why do i expect this bheavior:

`Second` is using the FallbackPolicy and `Third` is using the specified policy with the name `Default`.
Both the FallbackPolicy and the named policy `Default` use the same policy object.

### Steps To Reproduce

As already mentioned the problem can be reproduced with a WebApp project template.
Things changed on that template:
- HelloWorldController.cs file added with the HelloWorldController as shown above.
- Program.cs changed with the changes mentioned above.
  - Make sure to get a valid security key from method `GetSymmetricSecurityKey`.

[https://github.com/JanKotschenreuther/PolicyApiDemo.git](https://github.com/JanKotschenreuther/PolicyApiDemo.git)

1. Create a JWT which does not contain the required claim `SomeClaim_1`.
2. Send a Http-Get-Request to the api/HelloWorld/Second
    1. Add the JWT to the `Authorization` header with Authorization scheme `Bearer`
3. Send a Http-Get-Request to the api/HelloWorld/Third
    1. Add the JWT to the `Authorization` header with Authorization scheme `Bearer`
4. Compare responses, both should return the same Response with HTTP-StatusCode 403 and ReasonPhrase Forbidden.

### .NET Version

7.0.203

### Anything else?

TargetFramework of the project I have reproduced this problem with:
- net7.0

---

dotnet --info

.NET SDK:
 Version:   7.0.203
 Commit:    5b005c19f5

Laufzeitumgebung:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.203\

Host:
  Version:      7.0.5
  Architecture: x64
  Commit:       8042d61b17

.NET SDKs installed:
  6.1.426 [C:\Program Files\dotnet\sdk]
  7.0.101 [C:\Program Files\dotnet\sdk]
  5.0.104 [C:\Program Files\dotnet\sdk]
  5.0.214 [C:\Program Files\dotnet\sdk]
  5.0.303 [C:\Program Files\dotnet\sdk]
  5.0.401 [C:\Program Files\dotnet\sdk]
  5.0.408 [C:\Program Files\dotnet\sdk]
  8.0.202 [C:\Program Files\dotnet\sdk]
  6.0.203 [C:\Program Files\dotnet\sdk]
  6.0.311 [C:\Program Files\dotnet\sdk]
  6.0.400-preview.22330.6 [C:\Program Files\dotnet\sdk]
  9.0.203 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.19 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.19 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.19 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

**Comment by JanKotschenreuther** [comment]
Sorry for the missing steps to reproduce.
I edited the issue.
- Steps to reproduce added.
- Reference to a public GitHub repo which contains a minimalistic repro project added.

**Comment by JanKotschenreuther** [comment]
This is no bug.
I was able to find out what I did wrong here.

The FallbackPolicy only applies when no `IAuthorizeData` has been provided for an endpoint.
The `AuthorizeAttribute` implements `IAuthorizeData`.

The endpoint `Second` is not annotated with an `AuthorizeAttribute`.
Because of that, I thought that the FallbackPolicy has to apply.

What I did not recognize in the first place, was the fact that the extension method `RequireAuthorization()` adds a `new AuthorizeAttribute()` to every endpoint.
```csharp
app.UseEndpoints(endpoints =>
{
    endpoints.MapControllers().RequireAuthorization();
});
```

That means after `RequireAuthorization()` has been called, every endpoint is already provided with `IAuthorizeData` and thus the FallbackPolicy is not evaluated, which is the expected behavior.

So, if I want a policy to apply to all endpoints, which got no policy applied, there the following options:
- Use the FallbackPolicy. if `RequireAuthorization()` is NOT used.
- Use the DefaultPolicy, if `RequireAuthorization()` is used.
- Use `RequireAuthorization()` and pass policies, if they should be used on every endpoint.
    - The policy will be evaluated on an AND basis with other policies applied like with `AuthorizeAttribute`.

All this might be obvious, but maybe additional docs or highlighted notes in the docs can avoid other people to run into this pitfall.
